build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/transport/TransportTestExistTask.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.internal.precommit.transport;
+
+import org.elasticsearch.gradle.internal.conventions.precommit.PrecommitTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.tasks.CacheableTask;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.process.ExecOperations;
+import org.gradle.workers.WorkAction;
+import org.gradle.workers.WorkParameters;
+import org.gradle.workers.WorkQueue;
+import org.gradle.workers.WorkerExecutor;
+import org.objectweb.asm.ClassReader;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+/**
+ * Runs LoggerUsageCheck on a set of directories.
+ */
+@CacheableTask
+public abstract class TransportTestExistTask extends PrecommitTask {
+
+    private FileCollection classpath;
+
+    private final ListProperty<FileCollection> classesDirs;
+
+    private ObjectFactory objectFactory;
+    private FileCollection mainSources;
+    private FileCollection testSources;
+    private Configuration compileClasspath;
+    private Configuration testClasspath;
+
+    @Inject
+    public TransportTestExistTask(ObjectFactory objectFactory) {
+        this.classesDirs = objectFactory.listProperty(FileCollection.class);
+        this.objectFactory = objectFactory;
+        setDescription("Runs TransportTestExistTask on output directories of all source sets");
+    }
+
+    @Inject
+    abstract public WorkerExecutor getWorkerExecutor();
+
+    @TaskAction
+    public void runLoggerUsageTask() {
+        WorkQueue workQueue = getWorkerExecutor().noIsolation();
+        workQueue.submit(TransportTestExistWorkAction.class, parameters -> {
+            parameters.getMainSources().setFrom(mainSources);
+            parameters.getTestSources().setFrom(testSources);
+            parameters.getCompileClasspath().setFrom(compileClasspath);
+            parameters.getTestClasspath().setFrom(testClasspath);
+        });
+    }
+
```

Comment:
sometimes we use @Input/@Output.
when should I use this?

Reply:
These kind of annotations are used to provide gradle information to calculate if a task is up-to-date or not. In general if inputs havn't changed (input files and input properties) and the output of a task is still there (untampered) gradle considers a task up-to-date and skips executation.

Reply:
there should be input file annotations for `mainSources`, `testSources`, `compileClasspath`, `testClasspath`; mainSources and testSources could probably just annotated with @InputFiles and the other two with @Classpath